### PR TITLE
Add initial implementation of `suggest_float`.

### PR DESCRIPTION
**Pull request by himkt**
related #510 

This PR adds a new method `suggest_float`, which is a wrapper of `suggest_uniform` and `suggest_loguniform`.

### CODE REVIEW

File: tests/test_trial.py [hunk 1]
```diff
@@ -25,6 +25,25 @@
     [storages.InMemoryStorage, lambda: storages.RDBStorage('sqlite:///:memory:')])
 
 
+@parametrize_storage
+def test_check_distribution_suggest_float(storage_init_func):
```
Comment (himkt) [inline tests/test_trial.py]: I added the test to check if:

- `suggest_float(...)` and `suggest_uniform(...)` are the same
- `suggest_float(..., log=True)` and `suggest_loguniform(...)` are the same

**Comment by himkt** [comment]
Now I check if `suggest_float` and `suggest_uniform` (`suggest_loguniform`) are the same by following:

```
x1 = trial.suggest_float('x1', 10, 20)
x2 = trial.suggest_uniform('x1', 10, 20)  # looks like the result of suggest_float is cached

assert x1 == x2
```

I'm not sure this is good for testing this PR and I would really appreciate any feedback.

File: optuna/trial.py [hunk 2]
```diff
@@ -146,6 +157,72 @@ def _init_relative_params(self):
         self.relative_params = self.study.sampler.sample_relative(self.study, trial,
                                                                   self.relative_search_space)
 
+    def suggest_float(self, name, low, high, log=False):
```
Comment (toshihikoyanase) [inline optuna/trial.py]: Please add a description for `log` in docstring.

File: tests/test_trial.py (hunk 1 above)
Comment (toshihikoyanase) [reply tests/test_trial.py]: I think checking the compatibility with the existing interface is enough for the first implementation. 

And also, I think we need to check the case `low==high`.

File: optuna/trial.py [hunk 3]
```diff
@@ -146,6 +157,72 @@ def _init_relative_params(self):
...
+                    learning_rate_init = trial.suggest_float('learning_rate_init',
+                                                             1e-5, 1e-3, log=True)
+                    clf = MLPClassifier(hidden_layer_sizes=(100, 50), momentum=momentum,
+                                        learning_rate_init=learning_rate_init,
+                                        solver='sgd', random_state=0)
+                    clf.fit(X_train, y_train)
+
+                    return clf.score(X_test, y_test)
+
+                study = optuna.create_study()
```
Comment (toshihikoyanase) [inline optuna/trial.py]: Maybe this objective function should be maximized. Please set `direction='maximize'`.

File: optuna/trial.py (hunk 2 above)
Comment (himkt) [reply optuna/trial.py]: Added in df6a469.

File: tests/test_trial.py (hunk 1 above)
Comment (himkt) [reply tests/test_trial.py]: 34d60b1
I added the test for checking the case `low==high`.

File: optuna/trial.py (hunk 3 above)
Comment (himkt) [reply optuna/trial.py]: :pray: 4a4e8d2
I'll send another PR to fix the same mistakes in `suggest_uniform` and `suggest_loguniform`.

**Comment by himkt** [comment]
@toshihikoyanase I revised my PR, please take a look.

File: optuna/trial.py [hunk 4]
```diff
@@ -148,6 +159,82 @@ def _init_relative_params(self):
             self.study, trial, self.relative_search_space
         )
 
+    @experimental("1.2.0")
+    def suggest_float(self, name, low, high, log=False):
+        # type: (str, float, float, bool) -> float
+        """Wrapper method for ``suggest_uniform`` and ``suggest_loguniform``.
+
+        .. versionadded:: 1.2.0
```
Comment (sile) [inline optuna/trial.py]: The next release version will be `v1.3.0`.
```suggestion
        .. versionadded:: 1.3.0
```

File: optuna/trial.py [hunk 5]
```diff
@@ -12,18 +13,28 @@
...
+    ]
+
 
 class BaseTrial(object, metaclass=abc.ABCMeta):
     """Base class for trials.
 
     Note that this class is not supposed to be directly accessed by library users.
     """
 
+    def suggest_float(self, name, low, high, log):
```
Comment (sile) [inline optuna/trial.py]: How about specifying the default value of `log` argument?
```suggestion
    def suggest_float(self, name, low, high, log=False):
```

File: optuna/trial.py [hunk 6]
```diff
@@ -148,6 +160,82 @@ def _init_relative_params(self):
             self.study, trial, self.relative_search_space
         )
 
+    @experimental("1.2.0")
+    def suggest_float(self, name, low, high, *, log=False):
+        # type: (str, float, float, bool) -> float
+        """Wrapper method for ``suggest_uniform`` and ``suggest_loguniform``.
```
Comment (sile) [inline optuna/trial.py]: Personally, at the first line, I prefer writing a sentence that briefly describes the behavior of the target method (see below as an example): 
```
Suggest a value for the floating point parameter.

Note that this is a wrapper method for :func:`~optuna.trial.Trial.suggest_uniform` and  :func:`~optuna.trial.Trial.suggest_loguniform`.
```
However, this is not a strong opinion and the current doc seems good as well.

File: optuna/trial.py [hunk 7]
```diff
@@ -148,6 +160,82 @@ def _init_relative_params(self):
...
+            )  # type: FloatingPointDistributionType
+        else:
+            distribution = distributions.UniformDistribution(low=low, high=high)
+
+        self._check_distribution(name, distribution)
+
+        if low == high:
+            return self._set_new_param_or_get_existing(name, low, distribution)
+
+        return self._suggest(name, distribution)
```
Comment (sile) [inline optuna/trial.py]: I might have understood incorrectly, but it seems that this code block can be simplified as follows:
```suggestion
        if log:
            return self.suggest_loguniform(low, high)
        else:
            return self.suggest_uniform(low, high)
```

File: optuna/trial.py (hunk 6 above)
Comment (himkt) [reply optuna/trial.py]: Thank you for the suggestion and I agree with your idea.
Let me apply your feedback. (6ff1d40)

File: optuna/trial.py (hunk 7 above)
Comment (himkt) [reply optuna/trial.py]: You're right, thank you.

File: optuna/trial.py (hunk 5 above)
Comment (himkt) [reply optuna/trial.py]: The default values were introduced in 988bdf9.

**Comment by himkt** [comment]
@sile 
Thank you for the reviewing and I'm sorry for the late reply.
I update my PR and apply some feedbacks.
Could you please take a look?

File: optuna/trial.py [hunk 8]
```diff
@@ -148,6 +158,84 @@ def _init_relative_params(self):
...
+                from sklearn.model_selection import train_test_split
+                from sklearn.neural_network import MLPClassifier
+
+                np.random.seed(seed=0)
+                X = np.random.randn(200).reshape(-1, 1)
+                y = np.random.randint(0, 2, 200)
+                X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
+
+
+            ..testcode::
```
Comment (sile) [inline optuna/trial.py]: Maybe a typo:
```suggestion
            .. testcode::
```

File: optuna/trial.py [hunk 9]
```diff
@@ -148,6 +158,84 @@ def _init_relative_params(self):
...
+                A flag to sample the value from the log domain or not.
+                If ``log`` is true, the value is sampled from the range in the log domain.
+                Otherwise, the value is sampled from the range in the linear domain.
+                See also :func:`suggest_uniform` and :func:`suggest_loguniform`.
+
+        Returns:
+            A suggested float value.
+        """
+
+        if log:
```
Comment (sile) [inline optuna/trial.py]: The comment https://github.com/optuna/optuna/pull/1021/files#r396901171 was marked as resolved but it seems the suggestion hasn't been applied.

File: optuna/trial.py (hunk 8 above)
Comment (himkt) [reply optuna/trial.py]: Thank you for catching!

File: optuna/trial.py (hunk 9 above)
Comment (himkt) [reply optuna/trial.py]: It seems I have resolved this unintentionally, sorry.
Applied in 4669c1b.

Comment (sile) [reply optuna/trial.py]: Thanks!

**Comment by crcrpar** [comment]
> So, please add `@experimental` annotation to `suggest_float`.

Do we really have `suggest_float` decorated with `@experimental`?

**Comment by toshihikoyanase** [comment]
@crcrpar Thank you for pointing it out.
The current `suggest_float` employes keyword-only arguments (https://www.python.org/dev/peps/pep-3102/), and users have a little/no troubles even if we add new arguments. So, I changed my mind and declined my suggestion about `@experimental`.
I should have commented. I'm sorry for bothering you.

**Comment by crcrpar** [comment]
Gotcha. Thanks for the explanation. That sounds modern and great way. 🙂

**Pull request merged**